### DISCUSSION TITLE
should Emit require function body to be nativeOnly/jsNative, etc?

### DISCUSSION THREAD

**Discussion by joprice** [fable-compiler/Fable#3873] [Q&A]
I hit an issue where `Emit` ended up being used on a function that provided an implementation. This was the result of copy pasting a method signature that had an `Emit` attribute and swapping the body for a real implementation, but it made me wonder if this should be a compile time error:

```fsharp
open Fable.Core

// valid usage  - Emit + jsNative
[<Emit("x")>]
let x: int = jsNative

// Missing Emit - results in runtime error - any reason this can't be a compile time error as well?
let y: int = jsNative

// Does not cause an issue. Body of function is replaced with "x"
[<Emit("x")>]
let z: int = 1
```

Is there any reason these cases of misuse should be valid or aren't possible to detect?

**Comment by MangelMaxime** [comment]
Right now, this code emit a warning in your compilation

> ```fsharp
> // Missing Emit - results in runtime error - any reason this can't be a compile time error as well?
> let y: int = jsNative
> ```

```text
FABLE: jsNative is being compiled without replacement, this will fail at runtime.
```

Personally, I don't see a problem with making it an `error` because the generated code will fails.

Regarding the 3 case:

> ```fsharp
> // Does not cause an issue. Body of function is replaced with "x"
> [<Emit("x")>]
> let z: int = 1
> ```

I think it can make sense to generate a similar level error as for using `jsNative` without a replacement because this is the same problem just reverse. Using wrote a function but replaced it, so we don't really know what he was expected to run and we should ask him to clarify the situation.

My intuition is that it should be possible to detect this case, because we need to check that the body is not equal to `jsNative` or equivalent. And we have control over the AST so I think this is do-able, I am not sure if this is difficult or not.

@dbrattli @ncave What is your opinion in generating an error instead of a warning if `jsNative` is called without a replacement ?

**Comment by ncave** [comment]
IMO it makes sense, but it could potentially break when recompiling some existing bindings (from Nuget) where those warning were just ignored.
